src/threading.ts

```diff
@@ -0,0 +1,118 @@
+import IPy_Object = interfaces.IPy_Object;
+import Py_CodeObject = require('./codeobject');
+import Py_FuncObject = require('./funcobject');
+import opcodes = require('./opcodes');
+import optable = require('./optable');
+import Py_Cell = require('./cell');
+import Py_FrameObject = require('./frameobject');
+import enums = require('./enums');
+
+var maxMethodResumes: number = 10000,
+    // The number of method resumes until Doppio should yield again.
+    methodResumesLeft: number = maxMethodResumes,
+    // How responsive Ninia should aim to be, in milliseconds.
+    responsiveness: number = 1000,
+    // Used for the CMA.
+    numSamples: number = 1;
+
+class Thread{
+    // Current state of Thread
+    private status: enums.ThreadStatus = enums.ThreadStatus.NEW;
+    private stack: Py_FrameObject[] = [];
+
+    // Executes bytecode method calls
+    public run(): void {
+
+        var stack = this.stack,
+            startTime: number = (new Date()).getTime(),
+            endTime: number,
+            duration: number,
+            estMaxMethodResumes: number;
+
+        methodResumesLeft = maxMethodResumes;
+
+        if (this.status === enums.ThreadStatus.NEW) {
+           this.setStatus(enums.ThreadStatus.RUNNABLE);
+        }
+        // If methodResumes not exceeded, execute the Py_FrameObject
+        // else, reset the counter, suspend thread and resume thread using setImmediate
+        // Use cumulative moving average to calculate to estimate number of methodResumes in one second
+        while (this.status === enums.ThreadStatus.RUNNING && stack.length > 0) {
+            var bytecodeMethod: Py_FrameObject = stack[stack.length - 1];
+            // Execute python bytecode methods
+            bytecodeMethod.exec(this);          
+
+            // If no method resumes are left, yield to javascript event loop
+            if (--methodResumesLeft === 0) {
+                endTime = (new Date()).getTime();
+                duration = endTime - startTime;
+                // Estimated number of methods we can resume before needing to yield.
+                estMaxMethodResumes = Math.floor((maxMethodResumes / duration) * responsiveness);
+                // Update CMA.
+                maxMethodResumes = Math.floor((estMaxMethodResumes + numSamples * maxMethodResumes) / (numSamples + 1));
+                numSamples++;
+                // Yield.
+                this.setStatus(enums.ThreadStatus.ASYNC_WAITING);
+                setImmediate(() => { this.setStatus(enums.ThreadStatus.RUNNABLE); });
+            }
+        }
+
+        if (stack.length === 0) {
+            // This thread has finished!
+            this.setStatus(enums.ThreadStatus.TERMINATED);
+        }
+    }
+
+    // Change Thread status
+    public setStatus(status: enums.ThreadStatus): void {
+        // Transition RUNNABLE thread to RUNNING 
+        if (this.status === enums.ThreadStatus.NEW) {
+            this.status = status;
```

Comment:
I see why you did this now. My suggestion above in `run()` should address the issue that caused you to do this.